added MSTG-RESILIENCE-4 and 5 to iOS (#1583)
OWASP/owasp-mstg: Document/0x06j-Testing-Resiliency-Against-Reverse-Engineering.md

#### Detection Methods

You can detect popular reverse engineering tools that have been installed in an unmodified form by looking for associated application packages, files, processes, or other tool-specific modifications and artifacts. In the following examples, we'll discuss different ways to detect the Frida instrumentation framework, which is used extensively in this guide and also in the real world. Other tools, such as Cydia Substrate / Cycript, can be detected similarly. Note that injection/hooking/DBI (Dynamic Binary Instrumentation) tools can often be detected implicitly, through run time integrity checks, which are discussed below.

For instance, in its default configuration on a jailbroken device, Frida runs as frida-server. When you explicitly attach to a target app (e.g. via frida-trace or the Frida REPL), Frida injects a frida-agent into the memory of the app. Therefore, you may expect to find it there after attaching to the app (and not before). On Android you can verify this, by grepping for the string "frida" in the memory (`maps`) of the process ID in the `proc` directory.

> On iOS the `proc` directory is not available, but you can list the loaded dynamic libraries in an app with the function `_dyld_image_count`.

Reviewer: - Added "however" to link to the prev. paragraph.
- Just wondering if `_dyld_image_count` is a public or private API? Does the Apple allow this on Store apps?

```suggestion
However, on iOS the `proc` directory is not available, but you can list the loaded dynamic libraries in an app with the function `_dyld_image_count`.
```

Author: _dyld_image_count is used in the iOS Security Suite. And this library was already implemented in at least one public app that passed the app review process.

> The other method, which also works for non-jailbroken devices, consists of embedding a [frida-gadget](https://www.frida.re/docs/gadget/ "Frida Gadget") into the IPA and _forcing_ the app to load it as one of its native libraries.

Reviewer: Added explicit reference to the name of the Frida mode, as in 0x4c.
```suggestion
Frida may also run in the so-called embedded mode, which also works for non-jailbroken devices. It consists of embedding a [frida-gadget](https://www.frida.re/docs/gadget/ "Frida Gadget") into the IPA and _forcing_ the app to load it as one of its native libraries.
```

> The application's static content, including its ARM-compiled binary and its external libraries, are stored inside the [Application].app directory. If you inspect the content of the `/var/containers/Bundle/Application/[UUID]/[Application].app` directory, you'll find the embedded frida-gadget as FridaGadget.dylib.

Reviewer: - changed to "is" because it refers to "static content"
- changed to `< >` for consistency with the rest of the guide.

```suggestion
The application's static content, including its ARM-compiled binary and its external libraries, is stored inside the `<Application>.app` directory. If you inspect the content of the `/var/containers/Bundle/Application/<UUID>/<Application>.app` directory, you'll find the embedded frida-gadget as FridaGadget.dylib.
```

| **Checking For Open TCP Ports** | The frida-server process binds to TCP port 27042 by default. Testing whether this port is open is another method of detecting the daemon. | This method detects frida-server in its default mode, but the listening port can be changed via a command line argument, so bypassing this is very trivial. |
| **Checking For Ports Responding To D-Bus Auth** | `frida-server` uses the D-Bus protocol to communicate, so you can expect it to respond to D-Bus AUTH. Send a D-Bus AUTH message to every open port and check for an answer, hoping that `frida-server` will reveal itself. | This is a fairly robust method of detecting `frida-server`, but Frida offers alternative modes of operation that don't require frida-server. |

Please remember that this table is far from exhaustive. For example, two other possible detection mechanisms are:

> - [named pipes](https://en.wikipedia.org/wiki/Named_pipe "Named Pipes") (used by frida-server for external communication), or

Reviewer: would it be possible to get to the named pipes from iOS? I'm not sure as we cannot run any "console" command.

Author: Very good question. So you can do FIFO in iOS https://stackoverflow.com/questions/38231687/is-it-possible-to-create-and-use-fifo-file-in-ios-objc, so I assume named pipes are working on iOS...

- detecting [trampolines](https://en.wikipedia.org/wiki/Trampoline_%28computing%29 "Trampolines") (see "[Prevent bypassing of SSL certificate pinning in iOS applications](https://www.guardsquare.com/en/blog/iOS-SSL-certificate-pinning-bypassing "Prevent bypassing of SSL certificate pinning in iOS applications")" for further explanation and sample code for detection of trampolines in an iOS app)

> Both would _help_ to detect Substrate or Frida's Interceptor but, for example, won't be effective against Frida's Stalker. Each of these detection methods will depend on whether you're using a jailbroken device, the specific version of the jailbreak and method and/or the version of the tool itself. At the end, this is part of the cat and mouse game of protecting data being processed on an uncontrolled environment (the end user's device).

Reviewer: ```suggestion
Both would _help_ to detect Substrate or Frida's Interceptor but, for example, won't be effective against Frida's Stalker. Remember that the success of each of these detection methods will depend on whether you're using a jailbroken device, the specific version of the jailbreak and method and/or the version of the tool itself. At the end, this is part of the cat and mouse game of protecting data being processed on an uncontrolled environment (the end user's device).
```

> It is important to note that these controls are only increasing the complexity of the reverse engineering process. If used, the best approach is to combine the controls cleverly instead of using them individually. However, none of them can assure a 100% effectiveness, as the reverse engineer will always have full access to the device and will therefore always win! You also have to consider that integrating some of the controls into your app might increase the complexity of your app and even have an impact on its performance.

#### Effectiveness Assessment

> Launch the app with various reverse engineering tools and frameworks installed on your test device. Include at least the following: Frida, Cydia Substrate, Cycript, SSL KillSwitch and the Needle Agent.

Reviewer: ```suggestion
Launch the app with various reverse engineering tools and frameworks installed on your test device. Include at least the following: Frida, Cydia Substrate, Cycript, SSL Kill Switch and the Needle Agent.
```

The app should respond in some way to the presence of those tools. For example by:

- Alerting the user and asking for accepting liability.
- Preventing execution by gracefully terminating.
- Securely wiping any sensitive data stored on the device.
- Reporting to a backend server, e.g, for fraud detection.

Next, work on bypassing the detection of the reverse engineering tools and answer the following questions:

- Can the mechanisms be bypassed trivially (e.g., by hooking a single API function)?
- How difficult is identifying the anti reverse engineering code via static and dynamic analysis?
- Did you need to write custom code to disable the defenses? How much time did you need?
- What is your assessment of the difficulty of bypassing the mechanisms?

The following steps should guide you when bypassing detection of reverse engineering tools:

> 1. Patch the anti reverse engineering functionality. Disable the unwanted behavior by patching the binary through usage of Hooper or Ghidra.

Reviewer: Replaced with radare2/Cutter as Hopper is commercial.
```suggestion
1. Patch the anti reverse engineering functionality. Disable the unwanted behavior by patching the binary through usage of radare2/Cutter or Ghidra.
```

> 2. Use Frida or Cydia Substrate to hook file system APIs on the Objective-C /  Swift or native layers. Return a handle to the original file, not the modified file.

Reviewer: ```suggestion
2. Use Frida or Cydia Substrate to hook file system APIs on the Objective-C/Swift or native layers. Return a handle to the original file, not the modified file.
```

Refer to the chapter [Tampering and Reverse Engineering on iOS](0x06c-Reverse-Engineering-and-Tampering.md)" for examples of patching and code injection.

### Testing Emulator Detection (MSTG-RESILIENCE-5)

#### Overview

> In the context of anti-reversing, the goal of emulator detection is to increase the difficulty of running the app on an emulated device, which impedes some tools and techniques reverse engineers like to use. This increased difficulty forces the reverse engineer to defeat the emulator checks or utilize the physical device, thereby barring the access required for large-scale device analysis.

Reviewer: I'm not sure about the first sentence. What it is meant with the tools and techniques? Which ones?

What about this simpler version?

```suggestion
The goal of emulator detection is to increase the difficulty of running the app on an emulated device. This forces the reverse engineer to defeat the emulator checks or utilize the physical device, thereby barring the access required for large-scale device analysis.
```

> This is however not a concern on iOS. The only available simulator is the one that ships with Xcode. Simulator binaries are compiled to x86 code instead of ARM code and apps compiled for a real device (ARM architecture) don't run in the simulator. This makes the simulator useless for black box analysis and reverse engineering. See also the section [Testing on the iOS Simulator](0x06b-Basic-Security-Testing.md "Testing on the iOS Simulator") in the basic security testing chapter for further information.

Reviewer: ```suggestion
However, this is not a concern on iOS. The only available simulator is the one that ships with Xcode. Simulator binaries are compiled to x86 code instead of ARM code and apps compiled for a real device (ARM architecture) don't run in the simulator. This makes the simulator useless for black box analysis and reverse engineering. See also the section [Testing on the iOS Simulator](0x06b-Basic-Security-Testing.md "Testing on the iOS Simulator") in the basic security testing chapter for further information.
```

Reviewer: What about an emulator like Corellium? What could be done in this case? Even if it might _vanish_ soon, there might be alternatives (?).

Author: Unfortunately I do not have access to Corellium to test it and I couldn't find anyone writing about it publicly. Therefore I didn't mention it here. 
As Jeroen was writing in the MASVS ticket https://github.com/OWASP/owasp-masvs/issues/332, there are also QEMU based solutions, but this is far away from being usable in a penetration test at the moment. So the only thing we can mention at the moment is the iOS Simulator.

Reviewer: I've modified the suggestion to reflect and overcome this. What do you think? :)

- Did you need to write custom code to disable the defenses? How much time did you need?
- What is your assessment of the difficulty of bypassing the mechanisms??

### Testing Reverse Engineering Tools Detection (MSTG-RESILIENCE-4)

#### Overview

Reverse engineers use a lot of tools, frameworks, and apps, many of which you've encountered in this guide. Consequently, the presence of such tools on the device may indicate that the user is attempting to reverse engineer the app.

> Some reverse engineering tools can only run on a jailbroken device, force the app into debugging mode or start a service on the mobile phone. Different ways would need to be implemented in the mobile app to detect a reverse engineering attack and react to it, like terminating the app.

Author: see above

For instance, in its default configuration on a jailbroken  device, Frida runs as frida-server. When you explicitly attach to a target app (e.g. via frida-trace or the Frida REPL), Frida injects a frida-agent into the memory of the app. Therefore, you may expect to find it there after attaching to the app (and not before). On Android you can verify this, by grepping for the string "frida" in the memory (`maps`) of the process ID in the `proc` directory.
On iOS the `proc` directory is not available, but you can list the dynamic libraries in an app with the function `_dyld_image_count`.

The other method (which also works for non-jailbroken devices) consists of embedding a [frida-gadget](https://www.frida.re/docs/gadget/ "Frida Gadget") into the IPA and _forcing_ the app to load it as one of its native libraries.

In the application.app data directory the app stores the static content, as well as the application's ARM-compiled binary including the apps's libaries. If you inspect the app's directory in `/var/containers/Bundle/Application/[UUID]/[Application].app` you'll find the embedded frida-gadget as FridaGadget.dylib.

```bash

> Svens-iPhone:/var/containers/Bundle/Application/AC5DC1FD-3420-42F3-8CB5-E9D77C4B287A/SwiftSecurity.app/Frameworks root# ls -alh

Author: Removed. Thanks

-rw-r--r--  1 _installd _installd  70M Nov 16 06:37 FridaGadget.dylib
-rw-r--r--  1 _installd _installd 3.8M Nov 16 06:37 libswiftCore.dylib
-rw-r--r--  1 _installd _installd  71K Nov 16 06:37 libswiftCoreFoundation.dylib
-rw-r--r--  1 _installd _installd 136K Nov 16 06:38 libswiftCoreGraphics.dylib
-rw-r--r--  1 _installd _installd  99K Nov 16 06:37 libswiftDarwin.dylib
-rw-r--r--  1 _installd _installd 189K Nov 16 06:37 libswiftDispatch.dylib
-rw-r--r--  1 _installd _installd 1.9M Nov 16 06:38 libswiftFoundation.dylib
-rw-r--r--  1 _installd _installd  76K Nov 16 06:37 libswiftObjectiveC.dylib
```

Looking at these _traces_ that Frida _lefts behind_, you might already imagine that detecting those would be a trivial task. And actually, so trivial will be bypassing that detection. But things can get much more complicated. The following table shortly presents a set of some typical Frida detection methods and a short discussion on their effectiveness.

> Some of the following detection methods are implemented in the [iOS Security Suite](https://github.com/securing/IOSSecuritySuite "iOS Security Suite").

| Method | Description | Discussion |
| --- | --- | --- |

> | **Check Signature of an App** | One detection mechanism would be to verify the signature of the app. In order to embed the frida-gadget within the IPA, it would need to be repackaged and resigned. There is a library available in Objective-C called [iOS-App-Security-Class](https://github.com/karek314/iOS-App-Security-Class "iOS-App-Security-Class") that checks if the app is encrypted or not, by verifying `cryptid`. If this value is 0, the app was dumped (e.g. "clutched") and the Fairplay DRM encryption is not present anymore.  | This indicates that someone is trying to tamper and reverse engineer an app. There are limitations in such an implementation, as you cannot easily test if the check for `cryptid` is properly working. The app will only be encrypted by Apple after you submitted it to the App Store. |

Reviewer 2: ```suggestion
| **Check Signature of an App** | One detection mechanism would be to verify the signature of the app. In order to embed the frida-gadget within the IPA, it would need to be repackaged and resigned. There is a library available in Objective-C called [iOS-App-Security-Class](https://github.com/karek314/iOS-App-Security-Class "iOS-App-Security-Class") that checks if the app is encrypted or not, by verifying `cryptid`. If this value is 0, the app was dumped (e.g. "clutched") and the Fairplay DRM encryption is not present anymore.  | This indicates that someone is trying to tamper and reverse engineer the app. There are limitations in such an implementation, as you cannot easily test if the check for `cryptid` is properly working.  |
```
I removed that last part, as I think it's unnecessary. With enterprise certs, too, the app would be signed, right? So when is that ever an issue when preventing RE? (I can only think of "hard to test, since it will complain during dev", or maybe even when submitting it to the app store?) In any case, that line could use some more explanation.

Author: I was going through the commit history of the iOS-App-Security-Class and the last change to the project is 1 year ago, but the last code commit to the library is already over 2 years old and there is an open issue that was not addressed and fixed for almost a year now. This library is not maintained. 
I removed the whole row now, as there is no API offered by the SDK to validate the signature on iOS. Suggesting this "hack" is not a good control and you are right it's also hard to develop and I am not sure what happens during the app review process when using it.

Author: Thanks for pointing out

> | **Check The Environment For Related Artifacts**  | Artifacts can be packaged files, binaries, libraries, processes, and temporary files. For Frida, this could be the frida-server running in the target (jailbroken) system (the daemon responsible for exposing Frida over TCP) or the frida libraries loaded by the app. | Inspecting running services is not possible for an iOS app on a non-jailbroken device. The Swift method [CommandLine](https://developer.apple.com/documentation/swift/commandline "CommandLine") is not available on iOS to query for information about running processes, but there are inofficial ways, like using [NSTask](https://stackoverflow.com/a/56619466 "How can I run Command Line commands or tasks with Swift in iOS?"). Nevertheless when using this method, the app will be rejected during the App Store review process. There is no other public API available to query for running processes or execute system commands within an iOS App. Even if it would be possible, bypassing this would be as easy as just renaming the corresponding Frida artifact (frida-server/frida-gadget/frida-agent). Another way to detect Frida, would be to walk through the list of loaded libraries and check for suspicious ones (e.g. those including "frida" in their names), which can be done by using `_dyld_get_image_name`.

Reviewer 2: ```suggestion
| **Check The Environment For Related Artifacts**  | Artifacts can be packaged files, binaries, libraries, processes, and temporary files. For Frida, this could be the frida-server (the daemon responsible for exposing Frida over TCP) running in the target (jailbroken) system or the frida libraries loaded by the app. | Inspecting running services is not possible for an iOS app on a non-jailbroken device. The Swift method [CommandLine](https://developer.apple.com/documentation/swift/commandline "CommandLine") is not available on iOS to query for information about running processes, but there are unofficial ways, such as using [NSTask](https://stackoverflow.com/a/56619466 "How can I run Command Line commands or tasks with Swift in iOS?"). However, when using this method, the app will be rejected during the App Store review process. There is no other public API available to query for running processes or execute system commands within an iOS App. Even if it would be possible, bypassing this would be as easy as just renaming the corresponding Frida artifact (frida-server/frida-gadget/frida-agent). Another way to detect Frida, would be to walk through the list of loaded libraries and check for suspicious ones (e.g. those including "frida" in their names), which can be done by using `_dyld_get_image_name`.
```

> | **Checking For Open TCP Ports** | The frida-server process binds to TCP port 27042 by default. Check whether this port is open is another method of detecting the daemon. | This method detects frida-server in its default mode, but the listening port can be changed via a command line argument, so bypassing this is a little too trivial. |

Reviewer 2: ```suggestion
| **Checking For Open TCP Ports** | The frida-server process binds to TCP port 27042 by default. Testing whether this port is open is another method of detecting the daemon. | This method detects frida-server in its default mode, but the listening port can be changed via a command line argument, so bypassing this is very trivial. |
```

| **Checking For Ports Responding To D-Bus Auth** | `frida-server` uses the D-Bus protocol to communicate, so you can expect it to respond to D-Bus AUTH. Send a D-Bus AUTH message to every open port and check for an answer, hoping that `frida-server` will reveal itself. | This is a fairly robust method of detecting `frida-server`, but Frida offers alternative modes of operation that don't require frida-server. |

> Please remember that this table is far from exhaustive. We could also start talking about:

Reviewer 2: ```suggestion
Please remember that this table is far from exhaustive. For example, two other possible detection mechanisms are:
```

- [named pipes](https://en.wikipedia.org/wiki/Named_pipe "Named Pipes") (used by frida-server for external communication), or
- detecting [trampolines](https://en.wikipedia.org/wiki/Trampoline_%28computing%29 "Trampolines") (see "[Prevent bypassing of SSL certificate pinning in iOS applications](https://www.guardsquare.com/en/blog/iOS-SSL-certificate-pinning-bypassing "Prevent bypassing of SSL certificate pinning in iOS applications")" for further explanation and sample code for detection of trampolines in an iOS app)

> Both would _help_ detecting Substrate or Frida's Interceptor but, for example, won't be effective against Frida's Stalker; and many other, more or less, effective detection methods. Each of them will depend on whether you're using a jailbroken device, the specific version of the jailbreak and method and/or the version of the tool itself. At the end, this is part of the cat and mouse game of protecting data being processed on an untrusted environment (an app running in the user device).

Reviewer 2: ```suggestion
Both would _help_ to detect Substrate or Frida's Interceptor but, for example, won't be effective against Frida's Stalker. Each of these detection methods will depend on whether you're using a jailbroken device, the specific version of the jailbreak and method and/or the version of the tool itself. At the end, this is part of the cat and mouse game of protecting data being processed on an uncontrolled environment (the end user's device).
```
I think this is what you were trying to say.

> > It is important to note that these methods are just increasing the complexity for the reverse engineer. If being used, the best approach is to combine them cleverly instead of using them individually. However, none of them can assure a 100% effectiveness, remember that the reverse engineer always wins! You also have to consider that integrating some of them into your app might increase the complexity of your app as well as considerably mine its performance.

Reviewer 2: ```suggestion
> It is important to note that these controls are only increasing the complexity of the reverse engineering process. If used, the best approach is to combine the controls cleverly instead of using them individually. However, none of them can assure a 100% effectiveness, as the reverse engineer will always have full access to the device and will therefore always win! You also have to consider that integrating some of the controls into your app might increase the complexity of your app and even have an impact on its performance.
```

#### Effectiveness Assessment

> Launch the app with various reverse engineering tools and frameworks installed in your test device. Include at least the following: Frida, Cydia Substrate, Cycript, SSL KillSwitch and the Needle Agent.

Reviewer 2: ```suggestion
Launch the app with various reverse engineering tools and frameworks installed on your test device. Include at least the following: Frida, Cydia Substrate, Cycript, SSL KillSwitch and the Needle Agent.
```

- Did you need to write custom code to disable the defenses? How much time did you need?
- What is your assessment of the difficulty of bypassing the mechanisms?

The following steps should guide you when bypassing detection of reverse engineering tools:

1. Patch the anti reverse engineering functionality. Disable the unwanted behavior by patching the binary through usage of Hooper or Ghidra.
2. Use Frida or Cydia Substrate to hook file system APIs on the Objective-C /  Swift or native layers. Return a handle to the original file, not the modified file.

Refer to the chapter [Tampering and Reverse Engineering on iOS](0x06c-Reverse-Engineering-and-Tampering.md)" for examples of patching and code injection.

### Testing Emulator Detection (MSTG-RESILIENCE-5)

#### Overview

In the context of anti-reversing, the goal of emulator detection is to increase the difficulty of running the app on an emulated device, which impedes some tools and techniques reverse engineers like to use. This increased difficulty forces the reverse engineer to defeat the emulator checks or utilize the physical device, thereby barring the access required for large-scale device analysis.

> To make a long story short, this is not applicable for iOS apps. The only available simulator is the one that ships with Xcode. Simulator binaries are compiled to x86 code instead of ARM code and apps compiled for a real device (ARM architecture) don't run in the simulator. This makes the simulator useless for black box analysis and reverse engineering. See also the section [Testing on the iOS Simulator](0x06b-Basic-Security-Testing.md "Testing on the iOS Simulator") in the basic security testing chapter for further information.

Reviewer 2: ```suggestion
This is however not a concern on iOS. The only available simulator is the one that ships with Xcode. Simulator binaries are compiled to x86 code instead of ARM code and apps compiled for a real device (ARM architecture) don't run in the simulator. This makes the simulator useless for black box analysis and reverse engineering. See also the section [Testing on the iOS Simulator](0x06b-Basic-Security-Testing.md "Testing on the iOS Simulator") in the basic security testing chapter for further information.
```
(We can only make a long story short, if we have the long story)

Author: fair enough ;-)
